Adding data confidentially information (#1746)
dfinity/portal: docs/developer-docs/security/general-security-best-practices.md

Partial information on data that is stored in the subnet state tree will always leak. Therefore, data with a low-entropy value may entirely leak and be fully exposed, such as a Boolean value that can only be either "True" or "False". Leakage on data with a high-entropy is negligible. 

There are two types of user-related data that may be stored in the subnet state tree. The first is when a user sends an ingress message to a canister, the message and the response are both stored in the subnet state tree to be retrieved securely by the client. The ingress message should contain a high-entropy nonce that is implemented by the agent and typically not exposed to the user. The message response is determined by the canister and may not contain a high-entropy value. If the canister response consists of a low-entropy value, then the data may be leaked to users other than the ingress message sender. 

> The second type of user-related data is certified variables maintained by a canister that are also exposed through the subnet state tree. If a canister places low-entropy data into the state tree, then the data may leak to users who should not have access to that piece of data. For more information on the confidentiality of certified variables, please see the [IC interface spec documentation.](https://github.com/dfinity/interface-spec/blob/master/spec/index.md#request-read-state-http-read-state)

Reviewer: Certified variables are stored in the subnet state tree, but not exposed via read state requests (linked at the end of the paragraph). They could only be leaked by a malicious node which can anyways leak anything.

What is not mentioned in this section at all though are canister's private custom sections which are canister- or developer-related rather than user-related. Similarly to ingress message responses, low-entropy private custom sections could leak to unauthorized users.

Author: Thank you @mraszyk ! I've added information regarding canister private custom sections, and removed the link to the read state requests information.

#### Recommendation

Avoid test and dev code in production code whenever possible.

> ### Data confidentiality on the Internet Computer

Author: Thanks Robin, I had overlooked that heading. I've moved this section up above that heading now. 

I think this section fits best in this 'General best practices' page since it is more of a general piece of information about what developers can expect regarding default data confidentiality rather than a direct recommendation for canister data encryption.

#### Security concern

When storing data on the Internet Computer, there are two levels of data access.

1. Nodes are able to read all data that is stored on a subnet. This includes all messages sent to or from a canister, along with all data stored in a canister. This means a node could extract all data available to a canister. This will change with the implementation of TEE-based security for nodes.

> 2. End user clients can only access whatever data that nodes and canisters have made available to them. If a node does not misbehave and leak data, clients can only read the responses to ingress messages and queries that they have sent. The canister decides what data is exposed to the client. 

Reviewer 2: Maybe make this a bit more explicit as follows? 
```suggestion
2. End user clients can only access whatever data that nodes and canisters have made available to them. If the subnet's nodes do not misbehave and leak data, clients can only read the responses to ingress messages and queries that they have sent. The canister decides what data is exposed to the client. 
```

> Data that is stored in the subnet state tree will always leak. Therefore, data with a low-entropy value may entirely leak and be fully exposed, such as a Boolean value that can only be either "True" or "False". Data with a high-entropy is negligible. 

Reviewer 3: ```suggestion
Partial information on data that is stored in the subnet state tree will always leak. Therefore, data with a low-entropy value may entirely leak and be fully exposed, such as a Boolean value that can only be either "True" or "False". Data with a high-entropy is negligible. 
```

There are two types of user-related data that may be stored in the subnet state tree. The first is when a user sends an ingress message to a canister, the message and the response are both stored in the subnet state tree to be retrieved securely by the client. The ingress message should contain a high-entropy nonce that is implemented by the agent and typically not exposed to the user. The message response is determined by the canister and may not contain a high-entropy value. If the canister response consists of a low-entropy value, then the data may be leaked to users other than the ingress message sender. 

The second type of user-related data is certified variables maintained by a canister that are also exposed through the subnet state tree. If a canister places low-entropy data into the state tree, then the data may leak to users who should not have access to that piece of data. 

> #### Recommendation 

Reviewer 2: minor thing: would it make sense to also link to the Interface spec section that @mraszyk is writing [here](https://github.com/dfinity/interface-spec/pull/209/files)?

Author: Added a link to the IC spec
